[RFC] Travis: Add Clang MSan/TSan builds.

fwalch opened the pull request.

fwalch committed:
CMake: Add option for Clang MSan.

fwalch commented:
Huh, we actually have an error:

```
UMR in ioctl_common_pre at offset 0 inside [0x7fff9f3242bc, +4) 
==5657== WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x7fe10e697f42 in uv__nonblock /opt/neovim-deps/linux-x64/build/src/libuv/src/unix/core.c:482
    #1 0x7fe10e456f57 in stream_set_blocking /home/travis/build/neovim/neovim/src/nvim/os/stream.c:24
    #2 0x7fe10c23f3c0 in mch_exit /home/travis/build/neovim/neovim/src/nvim/os_unix.c:184
    #3 0x7fe10cd5b7a1 in command_line_scan /home/travis/build/neovim/neovim/src/nvim/main.c:904
    #4 0x7fe10cd48364 in main /home/travis/build/neovim/neovim/src/nvim/main.c:212
    #5 0x7fe10996176c (/lib/x86_64-linux-gnu/libc.so.6+0x2176c)
    #6 0x7fe10af8023c in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0xe823c)
  Uninitialized value was created by an allocation of 'fname' in the stack frame of function 'spell_delete_wordlist'
    #0 0x7fe10b5c3d10 in spell_delete_wordlist /home/travis/build/neovim/neovim/src/nvim/spell.c:4117
SUMMARY: MemorySanitizer: use-of-uninitialized-value /opt/neovim-deps/linux-x64/build/src/libuv/src/unix/core.c:482 uv__nonblock
```

Also, the ASAN build fails, don't know why:

```
/home/travis/build/neovim/neovim/src/nvim/os_unix.c:187:3: error: implicit
      declaration of function 'free_all_mem' is invalid in C99
      [-Werror,-Wimplicit-function-declaration]
  free_all_mem();
```

justinmk commented:
Very excited about this, thank you!

justinmk commented on .ci/clang.sh:
```diff
@@ -15,13 +20,14 @@ if [ "$TRAVIS_OS_NAME" = "linux" ]; then
 	export CC=/usr/local/clang-$clang_version/bin/clang
 	symbolizer=/usr/local/clang-$clang_version/bin/llvm-symbolizer
 	export ASAN_SYMBOLIZER_PATH=$symbolizer
+	export MSAN_SYMBOLIZER_PATH=$symbolizer
```
Won't the symbolizer path always be the same? So we could just rename `ASAN_SYMBOLIZER_PATH` to `SAN_SYMBOLIZER_PATH` or something like that.

fwalch replied:
I think the variable has to be named `ASAN_SYMBOLIZER_PATH` (same for MSAN) for Clang to pick it up.

http://clang.llvm.org/docs/MemorySanitizer.html

> Make sure that llvm-symbolizer binary is in PATH, or set environment variable MSAN_SYMBOLIZER_PATH to point to it.

Or we could maybe change the PATH, but looks like `$symbolizer` is used in some other variables as well, so we can't get rid of it. So rename `$symbolizer` to `SAN_SYMBOLIZER_PATH`?

justinmk replied:
No problem, I don't want to complicate anything.

justinmk commented on CMakeLists.txt:
```diff
@@ -191,13 +191,20 @@ include_directories(SYSTEM ${LIBTERMKEY_INCLUDE_DIRS})
+  message(ERROR "CLANG_ASAN and CLANG_MSAN cannot be enabled simultaneously.")
+endif()
+if((CLANG_ASAN OR CLANG_MSAN) AND NOT CMAKE_C_COMPILER_ID MATCHES "Clang")
+  message(WARNING "Sanitizers are only supported for Clang, disabling.")
```
Might as well make this an error.

fwalch replied:
:+1:

justinmk commented:
The MSan stacktrace is a little misleading, the actual bug appears to be in `spell_delete_wordlist()` which is ultimately called by `free_all_mem()`. Of course this problem wouldn't normally be caught since it happens right before exit. Nice catch by MSan (where was valgrind all this time??? :grin:).

justinmk committed:
MSan: Initialize buffer in spell_delete_wordlist.

justinmk commented:
https://github.com/neovim/neovim/pull/2811 seems to make MSan happy. Note that after tracing the code path in gdb, I verified that `vim_snprintf()` *does* initialize `fname` and nul-terminates it. So this really seems to be a minor quirk in MSan (or perhaps a configuration option we're missing).

fwalch commented:
I noticed that we didn't enable ThreadSanitizer (Undefined Behavior sanitizer was on); added it to the ASAN build for now to see if it works, but maybe it should be separate.

justinmk commented:
LGTM after comments. Feel free to cherry-pick https://github.com/neovim/neovim/pull/2811 and merge this and https://github.com/neovim/neovim/pull/2806

fwalch commented:
@justinmk Thanks! Will merge after the build passed.

fwalch committed:
CMake: Add option for Clang TSan, rename option ASAN -> ASAN_UBSAN.

fwalch commented:
Added a separate build type for TSAN.

fwalch commented:
Also, should `-DEXITFREE` be enabled for MSAN/TSAN?

fwalch committed:
CMake: Enable EXITFREE for all *San.

justinmk commented:
Can TSAN and/or UBSAN wait in a separate PR so other builds can start using MSAN?

fwalch commented:
> Can TSAN and/or UBSAN wait in a separate PR so other builds can start using MSAN?

Your patch fixed some MSan error that occured early, but now the functional tests won't run when it's enabled :-(

fwalch committed:
Travis: Use Clang 3.6.

justinmk commented:
> but the functional tests still don't run when it's enabled :-(

but the build in https://github.com/neovim/neovim/pull/2811 passed...?

fwalch commented:
> but the build in #2811 passed...?

You had only cherry-picked the Travis changes, not the ones in CMake (so *SAN didn't run at all).

fwalch commented:
Updated Clang. TSan works now, MSan still fails with SIGPIPE for the functional tests, and ASan/UBSan reports some new errors:

```
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
/home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97:26: runtime error: null pointer passed as argument 1, which is declared to never be null
/home/travis/build/neovim/neovim/build/include/memory.h.generated.h:13:114: note: nonnull attribute specified here
SUMMARY: AddressSanitizer: undefined-behavior /home/travis/build/neovim/neovim/src/nvim/msgpack_rpc/helpers.c:97 
Runtime errors detected
```

Pyrohh commented:
@fwalch 
> and ASan/UBSan reports some new errors

If you manage to fix that, please put "closes #2533" in the commit message.

fwalch committed:
Travis: Disable functional tests and allow failures for MSan.

 * Functional tests fail with SIGPIPE: disable them until we
   figure out the exact problem.
 * MSan reports some warnings: allow failures for the Travis
   build to allow fixing them in individual follow-up PRs.

justinmk commented on .ci/clang.sh:
```diff
@@ -38,11 +48,14 @@ build/bin/nvim --version
 # Run functional tests.
-if ! $MAKE_CMD test; then
+# FIXME (fwalch): Disabled for MSAN because of SIGPIPE error.
+if [ "$TRAVIS_OS_NAME" = "linux" -a "$CLANG_SANITIZER" -ne "MSAN" ]; then
```
I could be spreading FUD but I believe `[ ... ] && [ ... ]` is the most reliable "and" form in [ba]sh conditionals. I also thought `-ne` was for integer-likes, not strings[1]? I would use this:

    if [ "$TRAVIS_OS_NAME" = "linux" ] && [ "$CLANG_SANITIZER" -ne "MSAN" ]; then

[1] Note that `man test` does indicate that `-ne` is for integers.

fwalch replied:
:+1:

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -197,10 +197,12 @@ static bool os_stat(const char *name, uv_stat_t *statbuf)
-  *statbuf = request.statbuf;
-  uv_fs_req_cleanup(&request);
-  return (result == kLibuvSuccess);
+  if(uv_fs_stat(uv_default_loop(), &request, name, NULL) == kLibuvSuccess) {
```
Nice catch! I wonder what bugs this might fix :)

justinmk commented:
These are very promising results so far!

fwalch commented:
Hmm... any ideas? Apparently my commit didn't fix that.

```
==7426== WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x7f2581892326 in os_isdir /home/travis/build/neovim/neovim/src/nvim/os/fs.c:60:7
    #1 0x7f258194fbbc in vim_getenv /home/travis/build/neovim/neovim/src/nvim/os/env.c:483:12
    #2 0x7f25819446c1 in expand_env_esc /home/travis/build/neovim/neovim/src/nvim/os/env.c:269:25
    #3 0x7f257f728dae in option_expand /home/travis/build/neovim/neovim/src/nvim/option.c:3215:3
    #4 0x7f257f7193a6 in set_init_1 /home/travis/build/neovim/neovim/src/nvim/option.c:1899:11
    #5 0x7f25801ce210 in early_init /home/travis/build/neovim/neovim/src/nvim/main.c:173:3
    #6 0x7f25801cfb7c in main /home/travis/build/neovim/neovim/src/nvim/main.c:199:3
    #7 0x7f257ceb176c in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2176c)
    #8 0x7f257e4965c0 in _start (/home/travis/build/neovim/neovim/build/bin/nvim+0xae5c0)
  Uninitialized value was created by an allocation of 'request' in the stack frame of function 'os_stat'
    #0 0x7f2581897fd0 in os_stat /home/travis/build/neovim/neovim/src/nvim/os/fs.c:198
SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/travis/build/neovim/neovim/src/nvim/os/fs.c:60 os_isdir
```

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -192,13 +192,15 @@ int os_open(const char* path, int flags, int mode)
 static bool os_stat(const char *name, uv_stat_t *statbuf)
   FUNC_ATTR_NONNULL_ALL
 {
   uv_fs_t request;
```
This should make the error go away, although I wonder if it is covering up a deeper bug (perhaps in libuv?):

    uv_fs_t request = {0}

fwalch replied:
That works.

When [LibUV initializes the `uv_fs_t` data structure](https://github.com/libuv/libuv/blob/v1.x/src/unix/fs.c#L63-L70), it doesn't assign the `statbuf` field. So I guess it is a LibUV bug?

fwalch replied:
Okay, no, actually the problem is why we apparently return "true" even if there has been some problem and statbuf was not (fully) initialized.

justinmk replied:
But your change fixed that, right? See my other suggestion below.

fwalch replied:
You mean adding the `if`? No, that didn't fix it, which is why I think something else must be wrong (wrong `true` returned/access despite `false` returned?)

justinmk replied:
So does it make any difference if you init  `statbuf` in `os_getperm`:

    uv_stat_t statbuf = {0};

fwalch replied:
No, it doesn't.. so it's overwritten with the uninitialized values in `os_stat`. Also, now that I think of it, won't `= {0}` only initialize the first member of `uv_stat_t`?

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -192,13 +192,15 @@ int os_open(const char* path, int flags, int mode)
 /// Get stat information for a file.
 ///
-/// @return OK on success, FAIL if a failure occurred.
+/// @return `true` on success, `false otherwise.
```
missing a backtick

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -192,13 +192,15 @@ int os_open(const char* path, int flags, int mode)
   FUNC_ATTR_NONNULL_ALL
 {
   uv_fs_t request;
   int result = uv_fs_stat(uv_default_loop(), &request, name, NULL);
```
This might be a dumb idea, but it may be worth trying is to use a private loop instead of the default loop (see `stream_set_blocking()` for example). Just to see if MSAN changes its opinion.

justinmk commented:
Oh, I think it's complaining about the `const char *name` parameter, not `request`.

justinmk commented on src/nvim/os/fs.c:
```diff
@@ -192,13 +192,15 @@ int os_open(const char* path, int flags, int mode)
+  if (result == kLibuvSuccess) {
+    *statbuf = request.statbuf;
+  }
   uv_fs_req_cleanup(&request);
```
Hmm. I think it depends on what `uv_fs_req_cleanup(&request)` is doing. If `uv_fs_stat()` succeeds, we set `*statbuf = request.statbuf`. Instead I think we need to copy the buffer, not alias its pointer.

fwalch replied:
Commenting the cleanup does not make a difference :-/

fwalch commented:
No luck with the LibUV problem so far, but created a merge request for a problem detected with `libvterm`: https://code.launchpad.net/~fwalch/libvterm/initialize-colours-fixed/+merge/262455

```
==13149== WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x7f30341c6ac7 in kh_put_int_int_map /home/florian/Projects/neovim/src/nvim/map.c:114:1
    #1 0x7f30341cb42c in map_int_int_put /home/florian/Projects/neovim/src/nvim/map.c:114:1
    #2 0x7f3034c61785 in terminal_init /home/florian/Projects/neovim/src/nvim/terminal.c:164:5
    #3 0x7f30354aa0a9 in event_init /home/florian/Projects/neovim/src/nvim/os/event.c:60:3
    #4 0x7f3034a553a7 in main /home/florian/Projects/neovim/src/nvim/main.c:238:3
    #5 0x7f30327ca78f in __libc_start_main (/usr/lib/libc.so.6+0x2078f)
    #6 0x7f3033d4d7d8 in _start (/home/florian/Projects/neovim/build/bin/nvim+0x7f7d8)

  Uninitialized value was created by an allocation of 'color' in the stack frame of function 'terminal_init'
    #0 0x7f3034c61080 in terminal_init /home/florian/Projects/neovim/src/nvim/terminal.c:151

SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/florian/Projects/neovim/src/nvim/map.c:114 kh_put_int_int_map
```

fwalch commented:
Since fixing all the MSan errors might take a while, my plan for this PR is to set the MSan build to "allow failure" on Travis and then get this merged. Also the functional tests need to be fixed with MSan at some point. I will create issues for this once the PR is merged.

justinmk commented:
:+1:

Pyrohh commented:
> Since fixing all the MSan errors might take a while, my plan for this PR is to set the MSan build to "allow failure" on Travis and then get this merged

All right, although maybe https://github.com/fwalch/neovim/commit/12a765a2333864464739a75d24aeeaf80301d2f5 should probably be broken out into a separate PR? It's not clear that this PR contains it from the title IMO.

fwalch commented:
> It's not clear that this PR contains it from the title IMO. It might help get reviewers.

True.. but maybe I can convince @tarruda a quick look here :grin:  (since he wrote most of the surrounding code, I think): https://github.com/fwalch/neovim/commit/12a765a2333864464739a75d24aeeaf80301d2f5
@tarruda it's about fixing an ASan "passed NULL to parameter marked as NONNULL" warning in msgpack-rpc/API code (#2533).

fwalch commented:
Looking at coveralls, it only took 4 of the Travis jobs into account (clang ASAN, clang TSAN, OS X clang and OS X gcc). The last one, for whatever reason, also reports a coverage of 0%. Do we have to change the number of Travis jobs that coveralls has to wait for (AppVeyor has such a parameter) or something?

fwalch commented:
Followed @Pyrohh's suggestion and split out the fix into a different PR: #2925. Reordered some commits and rebased.

fwalch commented:
Functional tests with TSan are timing out ("no output received for 10 minutes") on Travis, but I can't reproduce this locally :-/

tarruda commented:
FWIW At this moment we have no multithreaded code, though libuv may use threads internally in some situations.

> Functional tests with TSan are timing out ("no output received for 10 minutes") on Travis, but I can't reproduce this locally :-/
 
Could this be caused by the natural slowness of a certain test boosted by travis and tsan? I recommend activating busted `gtest` output handler for more information on what test causes the hang.

tarruda commented:
@fwalch can you merge this without the tsan bits? I will add them on my upcoming PR which moves asynchronous IO to another thread.

fwalch commented:
@tarruda Okay, I'll remove the TSan Travis build (I assume the CMake changes are ok?).

tarruda commented:
> @tarruda Okay, I'll remove the TSan Travis build (I assume the CMake changes are ok?).

:+1:

tarruda commented on .ci/clang.sh:
```diff
@@ -2,26 +2,36 @@
 	export ASAN_SYMBOLIZER_PATH=$symbolizer
+	export MSAN_SYMBOLIZER_PATH=$symbolizer
 	export ASAN_OPTIONS="detect_leaks=1:log_path=$tmpdir/asan"
 	export TSAN_OPTIONS="external_symbolizer_path=$symbolizer:log_path=$tmpdir/tsan"
```
TSAN_OPTIONS uses a different syntax: https://code.google.com/p/thread-sanitizer/wiki/Flags

Here's the right setting:

```sh
export TSAN_OPTIONS="external_symbolizer_path=$symbolizer log_path=$tmpdir/tsan/tsan"
```

fwalch replied:
Thanks!

tarruda commented:
> @tarruda Okay, I'll remove the TSan Travis build (I assume the CMake changes are ok?).

Just leave the .travis.yml entry commented, I will uncomment in my PR when I fix the hang issue

fwalch committed:
Travis: Temporarily disable TSan build.

Tests currently time out and will be fixed by @tarruda in a future PR.

fwalch commented:
Fixed the TSAN_OPTIONS and commented the TSan build in the Travis config. The PR status is marked as "failed" because coveralls somehow only received input from one job (the clang ASAN one). The upload for the GCC build (which should not have changed in this PR), for example, failed with:

```
Traceback (most recent call last):
  File "/usr/local/bin/coveralls", line 9, in <module>
    load_entry_point('cpp-coveralls==0.3.9', 'console_scripts', 'coveralls')()
  File "/usr/local/lib/python2.7/dist-packages/cpp_coveralls/__init__.py", line 96, in run
    cov_report = coverage.collect(args)
  File "/usr/local/lib/python2.7/dist-packages/cpp_coveralls/coverage.py", line 375, in collect
    with io.open(source_file_path, mode='rb') as src_file:
IOError: [Errno 2] No such file or directory: '/home/travis/build/neovim/neovim/build/src/nvim/po/CMakeFiles/sjiscorr.dir/<built-in>'
coveralls upload failed.
```

The Travis build passed, though, so I think if someone can take a final look, this can be merged.

tarruda commented on .ci/clang.sh:
```diff
@@ -2,26 +2,36 @@
+	export MSAN_SYMBOLIZER_PATH=$symbolizer
 	export ASAN_OPTIONS="detect_leaks=1:log_path=$tmpdir/asan"
-	export TSAN_OPTIONS="external_symbolizer_path=$symbolizer:log_path=$tmpdir/tsan"
+	export TSAN_OPTIONS="external_symbolizer_path=$symbolizer log_path=$tmpdir/tsan"
```
It should be `log_path=$tmpdir/tsan/tsan`. Unlike with ASAN,  `log_path` doesn't specify the directory but the file pattern for the log.

fwalch replied:
I had a look at `asan_check` in `.ci/common.sh`, and I think this should be the file pattern for `check_logs` to find the log. I'll check again.

tarruda replied:
@fwalch that's correct, I don't know why I thought it should be the directory.

fwalch replied:
Thanks! Then I'll merge this.

tarruda commented:
Other than the last comment, LGTM. Feel free to merge anytime.
